lib/deimos/active_record_consume/batch_consumption.rb

```diff
@@ -88,36 +88,73 @@ def update_database(messages)
       def upsert_records(messages)
         key_cols = key_columns(messages)
 
-        # Create payloads with payload + key attributes
-        upserts = messages.map do |m|
-          attrs = if self.method(:record_attributes).parameters.size == 2
-                    record_attributes(m.payload, m.key)
-                  else
-                    record_attributes(m.payload)
-                  end
-
-          attrs&.merge(record_key(m.key))
-        end
-
+        # Create ActiveRecord Models with payload + key attributes
+        upserts = build_models(messages)
         # If overridden record_attributes indicated no record, skip
         upserts.compact!
+        # apply ActiveRecord validations and fetch valid Records
+        valid_upserts = filter_models(upserts)
+
+        save_records_to_database(@klass, key_cols, valid_upserts)
+
+        import_associations(valid_upserts, key_cols) unless @association_list.empty?
+      end
+
+      def save_records_to_database(record_class, key_cols, records)
+        columns = :all
+        if record_class.respond_to?(:bulk_import_columns)
+          columns = record_class.bulk_import_columns
+        end
 
         options = if key_cols.empty?
                     {} # Can't upsert with no key, just do regular insert
                   elsif ActiveRecord::Base.connection.adapter_name.downcase =~ /mysql/
                     {
-                      on_duplicate_key_update: :all
+                      on_duplicate_key_update: columns
                     }
                   else
                     {
                       on_duplicate_key_update: {
                         conflict_target: key_cols,
-                        columns: :all
+                        columns: columns
                       }
                     }
                   end
 
-        @klass.import!(upserts, options)
+        if @klass.respond_to?(:bulk_import_columns)
+          @klass.import!(columns, records, options)
+        else
+          @klass.import!(records, options)
+        end
+
+      end
+      # config for associations
+      # bulk_import_id is required only if handle_associations is set to true
+      # alternatively set a config to add which associations need to be considered
+      # There could be many associations not relevant to master table. For example, products and merchant association
+      def import_associations(entities)
+        associations = @klass.reflect_on_all_associations.select { |assoc| @association_list.include?(assoc.name) }
+
+        # create DB migration for bulk insert id. I did it manually in item-platform
+        table_by_bulk_import_id =
+          @klass.where(bulk_import_id: entities.map(&:bulk_import_id)).
```

Comment:
`bulk_import_id` is not required if they don't want to handle associations. it's something user has to define as a unique ID during build_models phase.
I am trying to understand how to prompt user to run a migration to create a column and index on this?

I will take a first pass on what I would do. 
Should I do something like `self.klass.columns.include?(@config.bulk_import_id_column)` in this method? 
If the column does not exist, raise an exception with information, say, "Create DB Migration on primary table and set bulk_import_id in build_models. Overriding build_models is mandatory if you want to handle associations"

Some message like this. 🤔

Reply:
Yeah something like that makes sense. I think we can treat `bulk_import_id` as - if you have it in your model, then we fully control it. So we can set it and read it back out. If you don't have it, but you indicated you want us to save your associations, then we can raise an error and say "please add this column by running `deimos install:bulk_import` or something similar.

Reply:
Done. This was tricky one. I ended up adding this to ActiveRecord class properties
e.g.,
```
class WidgetConsumer <  Deimos::ActiveRecordConsumer
          record_class Widget
          association_list :detail
          bulk_import_id_column :custom_id
end
```

I think this would not live in flipp_ruby_kafka kind of configuration level similar to record_class and association_list.

Reply:
I called the generator as `rails g deimos:bulk_import_id:setup`. Let me know if this is not a great choice for the generator. Will change the name.